kolibri/plugins/coach/assets/src/views/plan/LessonEditDetailsPage/index.vue

```diff
@@ -106,13 +108,58 @@
         }
         return this.$router.getRoute(route);
       },
+      initialAdHocLearners() {
+        return this.$store.state.adHocLearners.user_ids;
+      },
+      initialSelectedCollectionIds() {
+        let collectionIds = [];
+        // Only include the AdHocGroup in this if it has already
+        // had learners assigned to it.
+        this.lesson.lesson_assignments.forEach(assignment => {
+          if (assignment.collection_kind === CollectionKinds.ADHOCLEARNERSGROUP) {
+            if (this.hasAdHocLearnersAssigned) {
+              collectionIds.push(assignment.collection);
+            }
+          } else {
+            collectionIds.push(assignment.collection);
+          }
+        });
+        return collectionIds;
+      },
     },
     beforeRouteEnter(to, from, next) {
       return LessonResource.fetchModel({
         id: to.params.lessonId,
       })
         .then(lesson => {
-          next(vm => vm.setData(lesson));
+          next(vm => {
+            const collection = lesson.lesson_assignments.find(
+              a => a.collection_kind === CollectionKinds.ADHOCLEARNERSGROUP
+            );
+
+            if (collection) {
+              vm.$store
+                .dispatch('adHocLearners/initializeAdHocLearnersGroup', collection.collection)
+                .then(() => vm.setData(lesson));
+            } else {
+              // There is no 'ad hoc learners group' for this lesson, so we make one.
+              vm.$store
+                .dispatch('adHocLearners/createAdHocLearnersGroup', {
+                  classId: vm.$route.params.classId,
+                })
+                .then(() => {
+                  LessonResource.saveModel({
+                    id: vm.$route.params.lessonId,
+                    data: {
+                      lesson_assignments: [
+                        ...lesson.lesson_assignments,
+                        { collection: vm.$store.state.adHocLearners.id },
+                      ],
+                    },
+                  }).then(() => vm.setData(lesson));
```

Comment:
Instead of nesting this within a `.then`, you could just return the result of `LessonResource.saveModel` and chain this. I think that would be more the Promise-way